### PR TITLE
12.0 fix search with suspend security

### PR DESCRIPTION
**Pull request by mdpena**
Avoid Access Error when search with suspend security.

The Odoo `_search` function performs a  `sudo` on the first line of its
implementation, causing that the uid's wrapper (class BaseSuspendSecurityUid) is lost.

This is evidenced when a function with suspend security is called and in
its implementation needs the values ​​of a 'one2many' field (it does a
`search` in another model without reading access) an Access Error is raised.

### CODE REVIEW

File: base_suspend_security/models/base.py
Comment (hbrunn) [inline base_suspend_security/models/base.py]: please add a comment here that this is necessary because of https://github.com/OCA/OCB/blob/12.0/odoo/api.py#L858 which is used in https://github.com/OCA/OCB/blob/12.0/odoo/models.py#L4772

Comment (hbrunn) [reply base_suspend_security/models/base.py]: actually, this shouldn't be a problem, as we override __int__: https://github.com/OCA/server-backend/blob/8687422d0eed95a8e0cd2de1b39c621ab1e68c8b/base_suspend_security/base_suspend_security.py#L8 - can you debug where exactly the wrapper is lost?

Comment (mdpena) [reply base_suspend_security/models/base.py]: who has the wrapper is the self.uid, when doing sudo the `user` parameter not necessarily has the wrapper but still is important to maintain it and avoid not to break the thread.

Comment (mdpena) [reply base_suspend_security/models/base.py]: who has the wrapper is the self.uid, when doing sudo the `user` parameter not necessarily has the wrapper but still is important to maintain it and avoid to break the thread.
Also: 
```
>>> isinstance(int(BaseSuspendSecurityUid(1)), BaseSuspendSecurityUid)
False
```

In python 3.6:  DeprecationWarning: __int__ returned non-int (type BaseSuspendSecurityUid).  The ability to return an instance of a strict subclass of int is deprecated, and may be removed in a future version of Python.

Comment (hbrunn) [reply base_suspend_security/models/base.py]: ah, now I see the problem. We will get this issue in other places too, as some of the trickery of this module relies on being able to survive int(). But those we can handle as they come.

Now looking closer at this, I don't believe this is the correct patch. Something like

```python
env['some.model'].suspend_security().sudo(42)
```

must run as user 42 without suspended security. With your code, we'll never be able to unsuspend security.

Is there some other way to achieve what you want?

Comment (hbrunn) [reply base_suspend_security/models/base.py]: shouldn't we see the marker in `user` if this comes from a `suspend_security`'d search?

Comment (mvaled) [reply base_suspend_security/models/base.py]: > Now looking closer at this, I don't believe this is the correct patch. Something like
> 
> ```python
> env['some.model'].suspend_security().sudo(42)
> ```
> 
> must run as user 42 without suspended security. With your code, we'll never be able to unsuspend security.
> 
> Is there some other way to achieve what you want?

`BaseModel._search` does the following:

```python
        self.sudo(access_rights_uid or self._uid).check_access_rights('read')
```
If the sequence `suspend_security().sudo()` must reactivate the security, maybe the change is to test the argument to `sudo`:

```python
if isinstance(user, BaseSuspendSecurityUid):
   ...
```

Comment (mdpena) [reply base_suspend_security/models/base.py]: > ```python
> if isinstance(user, BaseSuspendSecurityUid):
> ```

I agree, that should work.

File: base_suspend_security/models/base.py
Comment (hbrunn) [inline base_suspend_security/models/base.py]: we don't need to wrap `user` again

File: base_suspend_security/tests/test_base_suspend_security.py
Comment (hbrunn) [inline base_suspend_security/tests/test_base_suspend_security.py]: I find this a little convoluted. Why not pick some specific model from the base module of which you know it's not accessible?

File: base_suspend_security/tests/test_base_suspend_security.py
Comment (hbrunn) [inline base_suspend_security/tests/test_base_suspend_security.py]: this line can go away

**Review by hbrunn — CHANGES_REQUESTED** [review]
I added a few nitpicks.

Do you agree we should remove the `__int__` implementation, as it doesn't do what we want anyways?

**Comment by mdpena** [comment]
Is done.

**Review by hbrunn — APPROVED** [review]
thanks!

**Comment by mvaled** [comment]
Hi @hbrunn 

Do we require another review?

**Comment by hbrunn** [comment]
yes

**Comment by mvaled** [comment]
> yes

Maybe @pedrobaeza could review this? ;)

**Comment by pedrobaeza** [comment]
Sorry, I'm out of the OCA board this year and stepped out from my community manager role, so you should look for another wildcard.

**Comment by mvaled** [comment]
> Sorry, I'm out of the OCA board this year and stepped out from my community manager role, so you should look for another wildcard.

Sorry to hear about that.  I hope everything goes well for you and your new endeavors.

**Comment by pedrobaeza** [comment]
> Sorry to hear about that. I hope everything goes well for you and your new endeavors.

Not any bad thing. Simply the phase has passed and now others have taken over.

**Comment by hbrunn** [comment]
@mvaled if you approve this, we can merge

**Review by mvaled — APPROVED** [review]

**Comment by mvaled** [comment]
I didn't know I could approve it.  I've just learnt how to do it.

**Comment by hbrunn** [comment]
@mvaled do you have a suggestion where to write about that so that more people are ware of this? It would be a pity to lose possible reviewers just because of this kind of misunderstandings.

/ocabot merge patch

**Comment by hbrunn** [comment]
please check https://travis-ci.org/OCA/server-backend/jobs/641058399#L331

**Comment by mvaled** [comment]
> @mvaled do you have a suggestion where to write about that so that more people are ware of this? It would be a pity to lose possible reviewers just because of this kind of misunderstandings.

Since I'm not a regular contributor to OCA projects I simply assumed I wasn't allowed.  I guess the README could point to general OCA guidelines about contributions and reviews.  Getting a formal statement of how do people get promoted from 'regular guy who reports and fixes some issues once in a while' to reviewer would help.  Specially for those seeking to get really involved the OCA, if they happen to land in any repository, a link to OCA's Get Involved page could help.

(I just noticed this https://odoo-community.org/page/review, so it seems anyone can submit a review. Cool!)

**Comment by hbrunn** [comment]
ah, then the omittance is mainly github stuff. you're allowed to review the hell out of all public repos here. rules here are we need to two approvals, but need someone with push permissions approve, and that would be me. and days in between, that's also agreed on for others to intervene if necessary.

but now the one thing that makes the merge bot fail can be fixed easily, and then we're done

**Comment by mvaled** [comment]
> but now the one thing that makes the merge bot fail can be fixed easily, and then we're done

I talked to @mdpena so that she works this out.

**Comment by mdpena** [comment]
@hbrunn @mvaled 

Is done. Thanks, both.

**Pull request merged**
